src/python/gudhi/random_point_generators.cc

```diff
@@ -21,48 +21,43 @@ namespace py = pybind11;
 typedef CGAL::Epick_d< CGAL::Dynamic_dimension_tag > Kern;
 
 template <typename Kernel>
-py::array_t<double> generate_points_on_sphere(py::object num_points, py::object dim, py::object radius) {
-    int npoints = num_points.cast<int>();
-    int d = dim.cast<int>();
-    double rad  = radius.cast<double>();
-    
-    py::gil_scoped_release release;
-
-    auto points_generated = Gudhi::generate_points_on_sphere_d<Kernel>(npoints, d, rad);
-    
-    py::gil_scoped_acquire acquire;
-    
-    py::array_t<double> points({npoints, d});
+py::array_t<double> generate_points_on_sphere(size_t num_points, int dim, double radius) {
+       
+    py::array_t<double> points({(int)num_points, dim});
  
     py::buffer_info buf = points.request();
-
     double *ptr = static_cast<double *>(buf.ptr);
 
-    assert(npoints == buf.shape[0]);
-    assert(d == buf.shape[1]);
+    assert(num_points == buf.shape[0]);
+    assert(dim == buf.shape[1]);
     
-
-    for (size_t i = 0; i < (size_t)npoints; i++)
-        for (size_t j = 0; j < (size_t)d; j++)
-            ptr[i*d+j] = points_generated.at(i).at(j);
+    std::vector<typename Kernel::Point_d> points_generated;
+    {
+        py::gil_scoped_release release;
+        points_generated = Gudhi::generate_points_on_sphere_d<Kernel>(num_points, dim, radius);
+        
+        for (size_t i = 0; i < num_points; i++)
+            for (size_t j = 0; j < (size_t)dim; j++)
+                ptr[i*dim+j] = points_generated[i][j];
+    }
 
     return points;
```

Comment:
I *think* it would be safe to include the return in the nogil region, because C++ implicitly moves `points` (not copy), and a move does not perform any python operation (it doesn't even touch the reference count like a copy would, although arguably since this is a new object there is no risk of concurrent access to it and touching the reference count should be safe). The advantage would be that you can get rid of the extra `{}` and can declare again points_generated where it is assigned, with `auto`.
But if the current version feels safer, I am ok with keeping it as is.